Refactor `write.rs`

ISSOtm opened the pull request:
*I'm aware this is a big PR, but I tried addressing that below the separation line.*

The change that motivated this rewrite was allowing `help` and `note` messages to show up even on label-less diagnostics. (This turned out to be very difficult to implement in the current architecture, so I figured it wouldn't hurt to clean it up as a way to learn its ins and outs.)

The other main upshot is that the printing logic should be far less opaque and hairy, and thus more maintainable. To the best of my knowledge, there is only one breaking change here, the new `Location` struct. It can be removed if compat breakage is to be avoided.

---

As for this PR's scope, I seem to understand from looking more broadly at this repo's issue tracker, that you'd like to spend less time on this crate; perhaps more on `ariadne2`, perhaps on something else (including non-programming).

I understand and relate to the sentiment, and I respect your decision whatever its rationale.
I would also like to improve ariadne further, as it's a great crate that I'd like to see some kinks of ironed out for my own projects.
I think I can do my part in bringing these two together by volunteering to help with maintenance. I am willing to start owning the same code I started hacking on and tweaking, and going through the issues and PRs.

Let me know if you're interested—I'm also available at my commit email address, if you'd rather discuss that topic privately.
I'll also understand if you decline that offer and/or reject this PR, of course :stuck_out_tongue:

ISSOtm committed:
Factor out a `nb_digits` function

And also use a simpler, `log10`-based logic

ISSOtm committed:
Dedicate a struct name to repeating chars

This makes its purpose clearer, and also removes one pair of parens

ISSOtm committed:
Factor out a `fetch_source` function

This ensures only a single copy of the error message exists,
and outlines yet more code outside of the massive
`write_for_stream` function.

ISSOtm committed:
Use `bool::then_some` to simplify a few exprs

ISSOtm committed:
Remove dead `impl Display for Show`

ISSOtm committed:
Mark `Show` and `Rept` as crate-local

This improves dead code analysis

ISSOtm committed:
Factor out a `max_line_no` function

Further deflating `write_for_stream`, though it also
should improve performance slightly:
- The number of digits is only computed for the greatest line number
- This is a valid transformation, because the largest line no is also
  one of the largest

ISSOtm committed:
Force `Show` to only contain an `Option`

Making its purpose clearer.

ISSOtm committed:
Use more compact print!/write! syntax

In the hopes of making the function a little shorter

ISSOtm committed:
Use the `format!` padding feature to pad out line numbers

This is more compact, and it might be cheaper too!

ISSOtm committed:
Factor out a function to obtain the appropriate `vbar`

ISSOtm committed:
Use a slightly more descriptive name for `is_line`

ISSOtm committed:
[BREAKING] Factor out a `Location` struct to group that info

The field names are more expressive than a tuple

ISSOtm committed:
Factor out a separate function for `group_connector`

To determine what goes to the left of a "source group" designation

ISSOtm committed:
Factor out a `is_referencing` method

It makes the purpose of the pointer comparison clearer
and is a little bit more concise.

ISSOtm committed:
Condense the arrow-character-selection logic

Factor out most of the repeated code

ISSOtm committed:
Extract the arrow bend selection logic into a separate function

ISSOtm committed:
Tweak collection of multi-line labels

Use `collect()` for hopefully better Vec prealloc,
derive one of the arrays from the other (since it's
a subset of the first), and sort the first array to
guarantee that the second one already is.

ISSOtm committed:
Elide one alloc per report emitted

`format_args!` is more difficult to use (hence why `code`
stays), but avoids allocating a `String`

ISSOtm committed:
Factor out a function for writing margin characters

ISSOtm committed:
Switch snapshot tests to Unicode

It has more character variety, and thus is more likely
to show differences that wouldn't be visible in ASCII

ISSOtm committed:
Add a few more tests

ISSOtm committed:
Build the `line_labels` array in a single statement

ISSOtm committed:
Factor out a function for checking if a label is the margin one

ISSOtm committed:
Avoid trying to render labels when printing help or notes

There aren't any labels left by that point

ISSOtm committed:
Split off `write_margin` into two functions

This eliminates a few parameters, and makes each "section"'s
purpose more evident.

Also stops trying to print arrows while printing note contents,
since there are no more labels at this point anyway.

ISSOtm committed:
Add multi-source tests

ISSOtm committed:
Factor out a function to write spacer lines

ISSOtm committed:
Print spacers before groups, rather than after

This leads to slightly simpler logic

ISSOtm committed:
Iterate on groups by ref instead of by value

Then, we can avoid holding onto the Vec's len across the whole loop

ISSOtm committed:
Rearrange a bool expr like Clippy suggests

ISSOtm committed:
Add a test for help and note but no labels

ISSOtm committed:
Do not apply the report kind's colour to its colon

For consistency with other crates, and `Note` and `Help`.

zesterer commented:
Wow, this is a great PR! I'm going to try to get some time to look through this properly, but unfortunately I'm travelling for 2 weeks as of tomorrow so it'll have to be after that. I really appreciate the time spent on this though!

ISSOtm commented:
No worries, then :)

As for the reviewing process, I *tried* making the commit log be easier to process piece-meal. (GitHub's UI allows you to review each commit's changes individually.) I think that should be all the more useful as the PR's cumulative changes are *really* large and complex.

Do let me know if you have doubts or questions about any part of this ^^

ISSOtm commented:
FWIW, this is still as much whitespace as the current code, there's just more labels so it's more whitespace.

I can look into removing trailing whitespace, but does it cause any practical trouble?

hellow554 commented:
Åh okay, it looked worse, sorry 🙈
Yes and no: I'm running ui tests on my program and often there's trailing white space that vscode likes to remove on every occasion.
Therefore I hoped that it can be "fixed"

ISSOtm commented:
Sublime Text has a "only trim trailing whitespace on lines I modified myself" option, which is more appropriate for this. I don't know if VS Code has something similar.

ISSOtm committed:
Ignore temp files generated by `cargo insta`

ISSOtm commented:
White rebasing this PR, I'm realising that there are several commits that could be cherry-picked into separate PRs. @zesterer, would you prefer if I did that?

ISSOtm commented:
I've rebased this PR against `main`, and ensured that every single commit passes `cargo test` on its own, so commits can be cherry-picked as necessary.

`fmt` and `clippy` fail, but they also do on `main`. Let me know if you'd like PRs that fix those.

nevakrien commented:
is this still being worked on? because i was considering tackling this problem myself

zesterer commented:
Urgh, rats. I'm so sorry: GitHub has, for some god-awful reason, decided to hide notifications for this issue from me until yesterday (and so, I'd forgotten that it existed). I guess this is what happens when Microsoft throws its entire dev budget into AI stock price pumps rather than actually building software.

@ISSOtm What state is this in? I think there have been quite a few upstream changes to this logic that might well conflict, sorry :(

ISSOtm commented:
I thought there wasn't interest in merging this, so I haven't taken the time to keep it current. Gimme a moment (two days, tops) and I should have it rebased, or at least an update on roadblocks.

ISSOtm commented:
> To avoid potential wasted work, did you have a quick synopsis of the intention behind this PR?

As the comment at the top of the file currently states,

https://github.com/zesterer/ariadne/blob/e526a3c5ea37942e96b8b656c834177d3dce0a60/src/write.rs#L9-L15

This was my attempt to learn it so I could make the changes I wanted to, and fixing the bugs I noticed along the way. I believe I can now say that I *do* understand it (both with and without this PR), and my intention was essentially to refactor the logic to simplify it.

> You mentioned help/note not working when no labels appear, but my gut feeling is that this shouldn't be too hard to bolt on to the existing codebase.

That was also my gut feeling, but then it turned out that the reason is pretty structural. I could have fixed it with fewer changes, by adding special cases and duplicating some printing code; however, this would have introduced more tech debt, which I prefer paying down when I can. (And I would've felt bad pushing tech debt onto *your* shoulders, heh.)

> `ariadne2` was intended to be conceptually cleaner and more modular, but I'm increasingly coming round to the view that it's feeling a bit over-engineered despite having nowhere near the functionality of `main` - so perhaps just iteratively improving `main` is the way to go instead.

I am better at improving existing designs than iterating on new ones, so I'm afraid there's little help I can give with `ariadne2`. I'm thinking that *if* we end up sharing maintainership, I could free up some of your free time and attention for you to work on it. It's fine to over-engineer it, probably, if it helps discover wrinkles in its implementation.

> Thanks for your original offer of maintenance: is this still something you're interested in?

Yes. As I mentioned above, I believe I now have significant understanding of the most complex part of `ariadne` (being `write.rs`).

Note that I will surely have to continue using my own fork for [my project](https://codeberg.org/ISSOtm/rsgbds) because I target a MSRV much lower than ariadne's current. But I guess maintaining the crate would make that job simpler for me? :P

zesterer commented:
> That was also my gut feeling, but then it turned out that the reason is pretty structural. I could have fixed it with fewer changes, by adding special cases and duplicating some printing code; however, this would have introduced more tech debt, which I prefer paying down when I can. (And I would've felt bad pushing tech debt onto _your_ shoulders, heh.)

I can imagine that the margin/gutter rendering code would need to be factored out, yes. Provided the upstream improvements are preserved, I'd be very happy to accept the refactored code!

> I am better at improving existing designs than iterating on new ones, so I'm afraid there's little help I can give with `ariadne2`. I'm thinking that _if_ we end up sharing maintainership, I could free up some of your free time and attention for you to work on it. It's fine to over-engineer it, probably, if it helps discover wrinkles in its implementation.

I'd be very grateful for that. I still enjoy working on the project (hence the recent burst of activity), but a few weeks ago I became a father - so the time I can dedicate to the project is intermittent, at best!

> Note that I will surely have to continue using my own fork for [my project](https://codeberg.org/ISSOtm/rsgbds) because I target a MSRV much lower than ariadne's current. But I guess maintaining the crate would make that job simpler for me? :P

Impressive work! I don't know much about the GB/GBC, but I've done quite a bit of [playing around with the GBA](https://www.youtube.com/watch?v=nS5rj80L-pk) of late: it's great to see new tools being built for old platforms.

I'll add you as maintainer tonight.

By the way: I've been considering a move to Codeberg (if only to escape the wave of AI slop on GitHub). I see you're quite active there: what are your thoughts about that?

ISSOtm committed:
Remove an unused function parameter

Shidoengie committed:
Split files into implementation and test modules (#5)

* Split write.rs and source.rs

* made generics more consistent

Shidoengie committed:
Split lib.rs along topics (#6)

zesterer reviewed (approved):
Approved, although there are two potential issues I noticed.

ISSOtm commented on src/report/write.rs:
```diff
@@ -0,0 +1,1000 @@
+        let groups = self.get_source_groups(&mut cache);
+
+        // Line number maximum width
+        let line_num_width = max_line_num(&groups).map_or(0, nb_digits);
```
You're absolutely corre\***cough**\* \***cough**\* Yeah, you're right. Probably some old leftover.

Shidoengie committed:
Fix incorrect double application of `nb_digits`

Fixes https://github.com/zesterer/ariadne/pull/119#discussion_r2931174045

ISSOtm commented on src/report/write.rs:
```diff
@@ -0,0 +1,1000 @@
+                .filter(|label_info| matches!(label_info.kind, LabelKind::Multiline))
+                .collect();
+            // Sort them by length; this also ensures that the next array is sorted.
+            multi_labels.sort_unstable_by_key(|label_info| !Span::len(&label_info.char_span));
```
AFAIK unstable sort is still deterministic, just not idempotent. I'd be surprised if it was non-deterministic, since it remains single-threaded behaviour?

ISSOtm committed:
Sort labels stably

Rationale discussed in
https://github.com/zesterer/ariadne/pull/119#discussion_r2931184643

ISSOtm commented:
(Note that I went with a merge commit rather than a rebase, so that the entirety of the PR can be seen as a whole. Squashing would've been an awful idea, too :P)
